src/menu.cc

```diff
@@ -105,106 +105,191 @@ extern "C" int nm_menu_hook(void *libnickel, nm_menu_item_t **items, size_t item
+            NM_LOG("...skipping action due to condition flags");
             continue;
+        }
+
+        free(err); // free the previous error if it is not NULL (so the last error can be saved for later)
+
+        nm_action_result_t *res = cur->act(cur->arg, &err);
+        if (err == NULL && res && res->type == NM_ACTION_RESULT_TYPE_SKIP) {
+            NM_LOG("...not updating success flag (value=%d) for skip result", success);
+        } else if (!(success = err == NULL)) {
+            NM_LOG("...error: '%s'", err);
+            continue;
+        } else if (!res) {
+            NM_LOG("...warning: you should have returned a result with type silent, not null, upon success");
+            continue;
+        }
+
+        NM_LOG("...result: type=%d msg='%s', handling...", res->type, res->msg);
+
+        MainWindowController *mwc;
+        switch (res->type) {
+        case NM_ACTION_RESULT_TYPE_SILENT:
+            break;
+        case NM_ACTION_RESULT_TYPE_MSG:
+            ConfirmationDialogFactory_showOKDialog(QString::fromUtf8(it->lbl), QLatin1String(res->msg));
+            break;
+        case NM_ACTION_RESULT_TYPE_TOAST:
+            mwc = MainWindowController_sharedInstance();
+            if (!mwc) {
+                NM_LOG("toast: could not get shared main window controller pointer");
+                break;
+            }
+            MainWindowController_toast(mwc, QLatin1String(res->msg), QStringLiteral(""), 1500);
+            break;
+        case NM_ACTION_RESULT_TYPE_SKIP:
+            skip = res->skip;
+            break;
+        }
+
+        if (skip == -1)
+            NM_LOG("...skipping remaining actions");
+        else if (skip != 0)
+            NM_LOG("...skipping next %d actions", skip);
+
+        nm_action_result_free(res);
+    }
+
+    if (err) {
+        NM_LOG("last action returned error %s", err);
+        ConfirmationDialogFactory_showOKDialog(QString::fromUtf8(it->lbl), QString::fromUtf8(err));
+        free(err);
+    }
+}
+
+QAction *AbstractNickelMenuController_createAction_before(QAction *before, nm_menu_location_t loc, bool last_in_group, void *_this, QMenu *menu, QWidget *widget, bool close, bool enabled, bool separator) {
+    int n = menu->actions().count();
+    QAction* action = AbstractNickelMenuController_createAction(_this, menu, widget, /*close*/false, enabled, /*separator*/false);
+
+    if (!menu->actions().contains(action)) {
+        NM_LOG("could not find added action at end of menu (note: old count is %d, new is %d), not moving it to the right spot or adding separator", n, menu->actions().count());
+        return action;
+    }
+
+    if (before != nullptr) {
+        menu->removeAction(action);
+        menu->insertAction(before, action);
+    }
 
-        NM_LOG("Adding item '%s'...", it->lbl);
+    if (close) {
+        // we can't use the signal which createAction can create, as it gets lost when moving the item
```

Comment:
That's why close & separator are forced false in the earlier `AbstractNickelMenuController_createAction`, right?

Reply:
Yes, exactly.

Reply:
^ I still don't know why the signal got lost when removing it, but I'd need a debug build of Qt to figure it out, and the SD card on my sacrificial Kobo failed last week.